src/OrleansDockerUtils/DockerMembershipOracle.cs

```diff
@@ -0,0 +1,468 @@
+                    // does not match the updated status.
+                    SiloEntry existing;
+                    if (!silos.TryGetValue(updatedSilo.SiloAddress, out existing))
+                    {
+                        // Add the new silo.
+                        if (added == null) added = new HashSet<SiloAddress>();
+                        added.Add(updatedSilo.SiloAddress);
+                        silos[updatedSilo.SiloAddress] = new SiloEntry(SiloStatus.Active, updatedSilo.Name)
+                        {
+                            Refreshed = true
+                        };
+                    }
+                    else
+                    {
+                        // Mark the existing silo as being refreshed.
+                        existing.Refreshed = true;
+                    }
+                }
+
+                // Identify removed silos.
+                foreach (var silo in silos)
+                {
+                    // Never remove self.
+                    if (silo.Key.Equals(SiloAddress)) continue;
+
+                    // Do not remove silos which were present in the update.
+                    if (silo.Value.Refreshed)
+                    {
+                        // Reset the refresh flag.
+                        silo.Value.Refreshed = false;
+                        continue;
+                    }
+
+                    // The silo was not present in the update, remove it.
+                    if (removed == null) removed = new HashSet<SiloAddress>();
+                    removed.Add(silo.Key);
+                }
+
+                // If any silos were removed, 
+                if (removed != null)
+                {
+                    foreach (var silo in removed)
+                    {
+                        silos.Remove(silo);
+                    }
+                }
+            }
+
+            // If anything was updated, clear the cache before notifying clients.
+            if (added != null || removed != null)
+            {
+                // Clear the caches.
+                ClearCaches();
+
+                // If any silos were added or removed, notify subscribers of the new status.
+                var siloStatusListeners = subscribers.Values.ToList();
+                if (added != null)
+                {
+                    foreach (var address in added)
+                    {
+                        NotifySubscribers(address, SiloStatus.Active, siloStatusListeners);
+                    }
+                }
+
+                if (removed != null)
+                {
+                    foreach (var address in removed)
+                    {
+                        NotifySubscribers(address, SiloStatus.Dead, siloStatusListeners);
+                    }
+                }
+            }
+        }
+
+        private async Task RefreshAsync()
+        {
+            try
+            {
+                await resolver.Refresh();
```

Comment:
Yeah you are right. I copy that over from the SF implementation. I just refactored the timer out of both gateway and oracle. Now it is inside the resolver and this dependency don't happen anymore. Thanks for notice that.